src/common/webrtc/api.rs

```diff
@@ -106,6 +110,53 @@ impl WebRtcSdp {
     }
 }
 
+struct AtomicSyncInner {
+    waker: AtomicWaker,
+    done: AtomicBool,
+}
+
+#[derive(Clone)]
+pub(crate) struct AtomicSync(Arc<AtomicSyncInner>);
+
+impl Default for AtomicSync {
+    fn default() -> Self {
+        Self(Arc::new(AtomicSyncInner {
+            waker: AtomicWaker::new(),
+            done: AtomicBool::new(false),
+        }))
+    }
+}
+
+impl AtomicSync {
+    pub(crate) fn done(&self) {
+        self.0
+            .done
+            .store(true, std::sync::atomic::Ordering::Relaxed);
+        self.0.waker.wake();
+    }
+    fn get(&self) -> bool {
+        self.0.done.load(std::sync::atomic::Ordering::Relaxed)
+    }
+}
+
+impl Future for AtomicSync {
+    type Output = ();
+    fn poll(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        if self.0.done.load(std::sync::atomic::Ordering::Relaxed) {
+            return Poll::Ready(());
+        }
+        self.0.waker.register(cx.waker());
+        if self.0.done.load(std::sync::atomic::Ordering::Relaxed) {
```

Comment:
do we need to repeat the done check on L52? would it not be more straightforward to use the else to register waker and return `Poll::Pending` right away?  Even if there's a possibility that a change happens after registering the waker, it would get caught on the next poll.

Reply:
Yes there could be a race condition where the waker was called before it was registered (a no-op) so the notification would be effectively lost. Also there is no point of registering a waker if the flag is updated already